Cleanup and document `-Z tls-model` (#71558)
rust-lang/rust: src/doc/unstable-book/src/compiler-flags/tls-model.md

# `tls_model`

The tracking issue for this feature is: None.

------------------------

Option `-Z tls-model` controls [TLS model](https://www.akkadia.org/drepper/tls.pdf) used to

> generate code for accessing `#[thread_local]` `static` items.

Reviewer: Technically speaking, the TLS model is supposed to be a per-variable attribute rather than a global one. However we don't currently have such an attribute, only a global setting which sets the default TLS model for all variables.

Supported values for this option are:

- `global-dynamic` - General Dynamic TLS Model (alternatively called Global Dynamic) is the most
general option usable in all circumstances, even if the TLS data is defined in a shared library
loaded at runtime and is accessed from code outside of that library.  
This is the default for most targets.

> - `local-dynamic` - model usable if the TLS data is defined in a shared library loaded at runtime
> and is accessed only from the same library, but not from other code.

Reviewer: ```suggestion
- `local-dynamic` - model usable if the TLS data is only accessed from the shared library or executable it is defined in.
```

Author: So, I wanted to say that it *may* be a shared library that *may* be loaded at runtime, but it may be an executable or something loaded at program startup as well.
I think I'll use your wording and add "The TLS data may be in a library loaded after startup (via `dlopen`)." in the end instead.

> - `initial-exec` - model usable if the TLS data is defined in a shared library loaded at load time
> (program startup), but not at runtime, and is accessed from code outside of that library.

Reviewer: ```suggestion
- `initial-exec` - model usable if the TLS data is defined the executable or in a shared library loaded at program startup. The TLS data must not be in a library loaded after startup (via `dlopen`).
```
